### PR TITLE
executor: fix data race in the AnalyzeColumnsExec

### LINKED ISSUE #32978: DATA RACE in the AnalyzeColumnsExec
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c0015e2e98 by goroutine 195:
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1939 +0x2cde
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1827 +0x396
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL.func1()
      /go/tidb/session/session.go:1787 +0x2da
  github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor()
      /go/tidb/session/session.go:1771 +0x358
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()
      /go/tidb/session/session.go:1775 +0x124
  github.com/pingcap/tidb/statistics/handle.(*Handle).execRestrictedSQLWithStatsVer()
      /go/tidb/statistics/handle/handle.go:141 +0x48f
  github.com/pingcap/tidb/statistics/handle.(*Handle).execAutoAnalyze()
      /go/tidb/statistics/handle/update.go:1193 +0xee
  github.com/pingcap/tidb/statistics/handle.(*Handle).autoAnalyzeTable()
      /go/tidb/statistics/handle/update.go:1103 +0x485
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleAutoAnalyze()
      /go/tidb/statistics/handle/update.go:1063 +0xb48
  github.com/pingcap/tidb/executor_test.testKillAutoAnalyze.func2()
      /go/tidb/executor/analyze_test.go:2731 +0x851
  github.com/pingcap/tidb/executor_test.testKillAutoAnalyze()
      /go/tidb/executor/analyze_test.go:2743 +0x804
  github.com/pingcap/tidb/executor_test.TestKillAutoAnalyzeV2()
      /go/tidb/executor/analyze_test.go:2752 +0x33
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
Previous read at 0x00c0015e2e98 by goroutine 235:
  github.com/pingcap/tidb/executor.(*AnalyzeColumnsExec).buildSubIndexJobForSpecialIndex()
      /go/tidb/executor/analyze.go:1214 +0x112
  github.com/pingcap/tidb/executor.(*AnalyzeColumnsExec).handleNDVForSpecialIndexes()
      /go/tidb/executor/analyze.go:1127 +0xce
  github.com/pingcap/tidb/executor.analyzeColumnsPushdown·dwrap·13()
      /go/tidb/executor/analyze.go:707 +0x71
Goroutine 195 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  github.com/pingcap/tidb/testkit/testmain.(*testingM).Run()
      /go/tidb/testkit/testmain/wrapper.go:28 +0x42
  go.uber.org/goleak.VerifyTestMain()
      /home/prow/go/pkg/mod/go.uber.org/goleak@v1.1.12/testmain.go:53 +0x59
  github.com/pingcap/tidb/executor_test.TestMain()
      /go/tidb/executor/main_test.go:69 +0x99c
  main.main()
      _testmain.go:1947 +0x264
Goroutine 235 (finished) created at:
  github.com/pingcap/tidb/executor.analyzeColumnsPushdown()
      /go/tidb/executor/analyze.go:707 +0x2394
  github.com/pingcap/tidb/executor.(*AnalyzeExec).analyzeWorker()
      /go/tidb/executor/analyze.go:363 +0x1f2
  github.com/pingcap/tidb/executor.(*AnalyzeExec).Next·dwrap·12()
      /go/tidb/executor/analyze.go:98 +0x64
==================
==================
WARNING: DATA RACE
Write at 0x00c0015e2d40 by goroutine 195:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).ClearStmtVars()
      /go/tidb/sessionctx/variable/session.go:1623 +0x2dad
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1942 +0x2d81
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1827 +0x396
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL.func1()
      /go/tidb/session/session.go:1787 +0x2da
  github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor()
      /go/tidb/session/session.go:1771 +0x358
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()
      /go/tidb/session/session.go:1775 +0x124
  github.com/pingcap/tidb/statistics/handle.(*Handle).execRestrictedSQLWithStatsVer()
      /go/tidb/statistics/handle/handle.go:141 +0x48f
  github.com/pingcap/tidb/statistics/handle.(*Handle).execAutoAnalyze()
      /go/tidb/statistics/handle/update.go:1193 +0xee
  github.com/pingcap/tidb/statistics/handle.(*Handle).autoAnalyzeTable()
      /go/tidb/statistics/handle/update.go:1103 +0x485
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleAutoAnalyze()
      /go/tidb/statistics/handle/update.go:1063 +0xb48
  github.com/pingcap/tidb/executor_test.testKillAutoAnalyze.func2()
      /go/tidb/executor/analyze_test.go:2731 +0x851
  github.com/pingcap/tidb/executor_test.testKillAutoAnalyze()
      /go/tidb/executor/analyze_test.go:2743 +0x804
  github.com/pingcap/tidb/executor_test.TestKillAutoAnalyzeV2()
      /go/tidb/executor/analyze_test.go:2752 +0x33
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
Previous read at 0x00c0015e2d40 by goroutine 235:
  github.com/pingcap/tidb/sessionctx/variable.(*SysVar).GetSessionFromHook()
      /go/tidb/sessionctx/variable/variable.go:178 +0x164
  github.com/pingcap/tidb/sessionctx/variable.GetSessionOrGlobalSystemVar()
      /go/tidb/sessionctx/variable/varsutil.go:191 +0x316
  github.com/pingcap/tidb/executor.getBuildStatsConcurrency()
      /go/tidb/executor/analyze.go:304 +0x4c
  github.com/pingcap/tidb/executor.(*AnalyzeColumnsExec).handleNDVForSpecialIndexes()
      /go/tidb/executor/analyze.go:1128 +0x104
  github.com/pingcap/tidb/executor.analyzeColumnsPushdown·dwrap·13()
      /go/tidb/executor/analyze.go:707 +0x71
Goroutine 195 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1504 +0x9d1
  github.com/pingcap/tidb/testkit/testmain.(*testingM).Run()
      /go/tidb/testkit/testmain/wrapper.go:28 +0x42
  go.uber.org/goleak.VerifyTestMain()
      /home/prow/go/pkg/mod/go.uber.org/goleak@v1.1.12/testmain.go:53 +0x59
  github.com/pingcap/tidb/executor_test.TestMain()
      /go/tidb/executor/main_test.go:69 +0x99c
  main.main()
      _testmain.go:1947 +0x264
Goroutine 235 (finished) created at:
  github.com/pingcap/tidb/executor.analyzeColumnsPushdown()
      /go/tidb/executor/analyze.go:707 +0x2394
  github.com/pingcap/tidb/executor.(*AnalyzeExec).analyzeWorker()
      /go/tidb/executor/analyze.go:363 +0x1f2
  github.com/pingcap/tidb/executor.(*AnalyzeExec).Next·dwrap·12()
      /go/tidb/executor/analyze.go:98 +0x64
================== 
```

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #32978

Problem Summary:

### What is changed and how it works?

- executor/analyze.go:707 it create a goroutine but it has not ```wg.Wait```. so it is a goroutine leak. then it will cause a data race.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -346,30 +350,32 @@ func (e *AnalyzeExec) analyzeWorker(taskCh <-chan *analyzeTask, resultsCh chan<-
...
+			case pkIncrementalTask:
+				resultsCh <- analyzePKIncremental(task.colIncrementalExec)
+			case idxIncrementalTask:
+				resultsCh <- analyzeIndexIncremental(task.idxIncrementalExec)
+			}
+		default:
+			select {
+			case <-exitCh:
+				return
+			default:
```
Comment (xuyifangreeneyes) [inline executor/analyze.go:374]: Can we use one select here?
```
for {
    select {
    case task := <- taskCh:
        ...
    case <-exitCh:
        return
    }
}
```

Comment (hawkingrei) [reply executor/analyze.go:374]: @xuyifangreeneyes  I can use one select If you are not careful about the lost data when exitch is closed.

Comment (xuyifangreeneyes) [reply executor/analyze.go:376]: Got it.

File: executor/analyze.go [hunk 2]
```diff
@@ -345,31 +350,32 @@ func (e *AnalyzeExec) analyzeWorker(taskCh <-chan *analyzeTask, resultsCh chan<-
...
-		case idxTask:
-			resultsCh <- analyzeIndexPushdown(task.idxExec)
-		case fastTask:
-			resultsCh <- analyzeFastExec(task.fastExec)
-		case pkIncrementalTask:
-			resultsCh <- analyzePKIncremental(task.colIncrementalExec)
-		case idxIncrementalTask:
-			resultsCh <- analyzeIndexIncremental(task.idxIncrementalExec)
+		select {
+		case task := <-taskCh:
```
Comment (xuyifangreeneyes) [inline executor/analyze.go:356]: Can we make worker exit when it finds `taskCh` is closed?

File: executor/analyze.go [hunk 3]
```diff
@@ -345,31 +350,32 @@ func (e *AnalyzeExec) analyzeWorker(taskCh <-chan *analyzeTask, resultsCh chan<-
...
+			case fastTask:
+				resultsCh <- analyzeFastExec(task.fastExec)
+			case pkIncrementalTask:
+				resultsCh <- analyzePKIncremental(task.colIncrementalExec)
+			case idxIncrementalTask:
+				resultsCh <- analyzeIndexIncremental(task.idxIncrementalExec)
+			}
+		default:
+			select {
+			case <-exitCh:
```
Comment (winoros) [inline executor/analyze.go:372]: This exitCh is closed just after we put all tasks into the taskCh. When will the case that `we lost data when exitCh is closed` would happen?

Comment (hawkingrei) [reply executor/analyze.go:372]:
```
select{
   case <-task:
   case <-exitch:
}
```

if ```exitch``` is closed and it still has the task in the task channel, ```select``` maybe run the ```exitch``` branch and  ignore the ones in the task channel.

Comment (hawkingrei) [reply executor/analyze.go]: removed

File: executor/analyze.go (hunk 2 above)
Comment (hawkingrei) [reply executor/analyze.go]: removed

**Comment by hawkingrei** [comment]
@xuyifangreeneyes @winoros PTAL

**Comment by hawkingrei** [comment]
@winoros @xuyifangreeneyes PTAL

**Review by winoros — APPROVED** [review]

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**